src/WebJobs.Script/Host/AggregateFunctionMetadataProvider.cs

```diff
@@ -17,44 +18,65 @@
         private ImmutableArray<FunctionMetadata> _functions;
+        private IFunctionMetadataProvider _hostFunctionMetadataProvider;
 
-        public WorkerFunctionMetadataProvider(ILogger<WorkerFunctionMetadataProvider> logger, IFunctionInvocationDispatcher invocationDispatcher)
+        public AggregateFunctionMetadataProvider(
+            ILogger<AggregateFunctionMetadataProvider> logger,
+            IFunctionInvocationDispatcher invocationDispatcher,
+            IFunctionMetadataProvider hostFunctionMetadataProvider)
         {
             _logger = logger;
-            _dispatcher = invocationDispatcher ?? throw new ArgumentNullException(nameof(invocationDispatcher));
+            _dispatcher = invocationDispatcher;
+            _hostFunctionMetadataProvider = hostFunctionMetadataProvider;
         }
 
         public ImmutableDictionary<string, ImmutableArray<string>> FunctionErrors
            => _functionErrors.ToImmutableDictionary(kvp => kvp.Key, kvp => kvp.Value.ToImmutableArray());
 
-        public async Task<ImmutableArray<FunctionMetadata>> GetFunctionMetadataAsync(IEnumerable<RpcWorkerConfig> workerConfigs, bool forceRefresh)
+        public async Task<ImmutableArray<FunctionMetadata>> GetFunctionMetadataAsync(IEnumerable<RpcWorkerConfig> workerConfigs, IEnvironment environment, bool forceRefresh)
         {
             IEnumerable<RawFunctionMetadata> rawFunctions = new List<RawFunctionMetadata>();
             IEnumerable<FunctionMetadata> functions = new List<FunctionMetadata>();
             _logger.FunctionMetadataProviderParsingFunctions();
+
+            bool workerIndexing = Utility.CanWorkerIndex(workerConfigs, environment);
+
             if (_functions.IsDefaultOrEmpty || forceRefresh)
             {
-                if (_dispatcher != null)
+                if (workerIndexing)
                 {
+                    if (_dispatcher == null)
+                    {
+                        throw new ArgumentNullException(nameof(_dispatcher));
+                    }
+
                     // start up GRPC channels
                     await _dispatcher.InitializeAsync(new List<FunctionMetadata>());
 
                     // get function metadata from worker, then validate it
                     rawFunctions = await _dispatcher.GetWorkerMetadata();
-                    functions = ValidateMetadata(rawFunctions);
 
-                    // set up invocation buffers and send load requests
-                    await _dispatcher.FinishInitialization(functions);
+                    if (!IsNullOrEmpty(rawFunctions))
+                    {
+                        functions = ValidateMetadata(rawFunctions);
+                        // set up invocation buffers and send load requests
+                        await _dispatcher.FinishInitialization(functions);
+                    }
+                }
+                if (IsNullOrEmpty(rawFunctions) || !workerIndexing)
```

Comment:
empty rawFunctions does not mean that the worker rejected indexing. We need a definite indicator / evaluation of a field that explicitly indicates that the worker does not identify this as a valid payload and cannot index.

Reply:
updated the code here.